File: grind75/cpp/week3/009-Longest-Repeating-Character-Replacement.cpp
// problem link: https://leetcode.com/problems/longest-repeating-character-replacement/description/

class Solution {
public:
    int characterReplacement(std::string s, int k) {
        std::unordered_map<char, int> count;
        int res = 0, l = 0;

        for (int r = 0; r < s.size(); ++r) {
            count[s[r]]++;

            // Compute max frequency value explicitly (less optimized)
            int maxFreq = 0;
            for (const auto& pair : count) {
                maxFreq = std::max(maxFreq, pair.second);
            }

            while ((r - l + 1) - maxFreq > k) {
                count[s[l]]--;
                l++;
            }

            res = std::max(res, r - l + 1);
        }
        return res;
    }
};


// optimized

class Solution {
public:
    int characterReplacement(std::string s, int k) {
        std::unordered_map<char, int> count;
        int res = 0, l = 0, maxFreq = 0;

        for (int r = 0; r < s.size(); ++r) {
            count[s[r]]++;
            maxFreq = std::max(maxFreq, count[s[r]]);

            while ((r - l + 1) - maxFreq > k) {
                count[s[l]]--;
                l++;
            }

            res = std::max(res, r - l + 1);
        }
        return res;
    }
};

File: grind75/week3/009-Longest-Repeating-Character-Replacement/009-Longest-Repeating-Character-Replacement.py
# problem link: https://leetcode.com/problems/longest-repeating-character-replacement/
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        res = 0
        l = r = 0
        
        for r in range(len(s)):
            count[s[r]] = 1 + count.get(s[r], 0)
            while (r - l + 1) - max(count.values()) > k:
                count[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res


# optimized

class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        count = {}
        res = 0
        l = r = maxF = 0
        
        for r in range(len(s)):
            count[s[r]] = 1 + count.get(s[r], 0)
            maxF = max(maxF, count[s[r]])
            while (r - l + 1) - maxF > k:
                count[s[l]] -= 1
                l += 1
            res = max(res, r - l + 1)
        return res

